src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,317 @@
+                {
+                    _eventQueue.Add(() =>
+                    {
+                        RemoveLRUs();
+                    });
+                }
+                // add the new node
+                var newCacheItem = new LRUCacheItem<TKey, TValue>(key, value, expirationTime);
+                _eventQueue.Add(() =>
+                {
+                    // Add a remove operation in case two threads are trying to add the same value. Only the second remove will succeed in this case.
+                    _doubleLinkedList.Remove(newCacheItem);
+                    _doubleLinkedList.AddFirst(newCacheItem);
+                });
+                _map[key] = newCacheItem;
+            }
+
+            return true;
+        }
+
+        /// Each time a node gets accessed, it gets moved to the beginning (head) of the list.
+        public bool TryGetValue(TKey key, out TValue value)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (!_map.ContainsKey(key))
+            {
+                value = default;
+                return false;
+            }
+
+            // make sure node hasn't been removed by a different thread
+            if (_map.TryGetValue(key, out var cacheItem))
+                _eventQueue.Add(() =>
+                {
+                    _doubleLinkedList.Remove(cacheItem);
+                    _doubleLinkedList.AddFirst(cacheItem);
+                });
+
+            value = cacheItem != null ? cacheItem.Value : default;
+            return cacheItem != null;
+        }
+
+        /// Removes a particular key from the cache.
+        public bool TryRemove(TKey key, out TValue value)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (!_map.TryGetValue(key, out var cacheItem))
+            {
+                value = default;
+                return false;
+            }
+
+            value = cacheItem.Value;
+            _eventQueue.Add(() => _doubleLinkedList.Remove(cacheItem));
+            return _map.TryRemove(key, out _);
+        }
+
+        /// <summary>
+        /// FOR TESTING ONLY.
+        /// </summary>
+        /// <returns></returns>
+        internal LinkedList<LRUCacheItem<TKey, TValue>> LinkedList => _doubleLinkedList;
+
+        /// <summary>
+        /// FOR TESTING ONLY.
+        /// </summary>
+        public long LinkedListCount => _doubleLinkedList.Count;
+
+        /// <summary>
+        /// FOR TESTING ONLY.
+        /// </summary>
+        public long MapCount => _map.Count;
+
+        /// <summary>
+        /// FOR TESTING ONLY.
+        /// </summary>
+        /// <returns></returns>
+        internal ICollection<LRUCacheItem<TKey, TValue>> MapValues => _map.Values;
+
+        /// <summary>
+        /// FOR TESTING ONLY.
+        /// </summary>
+        public long EventQueueCount => _eventQueue.Count;
```

Comment:
Some of these properties need to remain public until the PR is ready to be merged in because of the performance tests.

Reply:
What could we do to allow the performance test to continue to be runnable?

Reply:
@brentschmaltz Could the performance tests be modified to rely on internal rather than public properties?